### PR TITLE
[stdlib] Support negative indexing and checking bounds for VariadicList.

### PR DESCRIPTION
**Pull request by northstreet12**
Related to issue https://github.com/modularml/mojo/issues/2864

### CODE REVIEW

**Review by gabrieldemarmiesse — COMMENTED** [review]
This is a nice quality of life feature. Thanks! I think we can make the code slighly simpler, can you investigate this?

File: stdlib/src/builtin/builtin_list.mojo [hunk 1]
```diff
@@ -157,7 +157,15 @@ struct VariadicList[type: AnyTrivialRegType](Sized):
...
             The element on the list corresponding to the given index.
         """
-        return __mlir_op.`pop.variadic.get`(self.value, idx.value)
+        var normalized_idx = idx
+        debug_assert(
+            -len(self) <= normalized_idx < len(self),
+            "Index must be within bounds",
+        )
+        if normalized_idx < 0:
+            normalized_idx += len(self)
```
[the remark this answers is not in the document]
Comment (northstreet12) [reply stdlib/src/builtin/builtin_list.mojo]: Yes, that is great! I missed it before. Thanks.

**Review by JoeLoser — COMMENTED** [review]
This is good.  Should we also do the same for `VariadicPack` and `VariadicListMem`?  The asymmetry seems weird to me.  It seems we should normalize the indices for all of them or none of them.  😄

**Comment by northstreet12** [comment]
> This is good. Should we also do the same for `VariadicPack` and `VariadicListMem`? The asymmetry seems weird to me. It seems we should normalize the indices for all of them or none of them. 😄

Fair enough. Fixed VariadicListMem the same way. But it seems somthing different for `VariadicPack`. I tried to use `@parameter` function to normalize index for `VariadicPack` then problems appeared. Should we modify this interface first?
```
    fn __getitem__[
        index: Int
    ](self) -> ref [Self.lifetime] element_types[index.value]:
```
Or some other sugesstions?

**Comment by gabrieldemarmiesse** [comment]
@northstreet12 can you give what code you tried and the error message your got for `VariadicPack`? I'll see if I can help you there.

**Comment by northstreet12** [comment]
> @northstreet12 can you give what code you tried and the error message your got for `VariadicPack`? I'll see if I can help you there.

I tried to add a  `@parameter` function like this:
```
    @parameter
    fn parameter_index_normalize[idx: Int, container_length: Int]() -> Int:
        return idx + int(idx < 0) * container_length
```
but this interface uses the value `index` at compile-time
```
    @always_inline
    fn __getitem__[
        index: Int
    ](self) -> ref [Self.lifetime] element_types[index.value]:
```
It seems no chance to involk the `@parameter` function.
How should we change the value of `index` at compile-time?

**Comment by gabrieldemarmiesse** [comment]
@northstreet12 I see what you mean. I tried it myself and it's quite hard to do. @JoeLoser can we skip `VariadicPack` for this PR, it requires quite a few changes to other parts of the codebase. Due to the complexity, I believe this deserves its own pull request. Would that be all right?

**Comment by northstreet12** [comment]
@JoeLoser Hi, any comments?

**Review by JoeLoser — COMMENTED** [review]
The changes are correct/fine.  This may affect performance given how widely variadics are used and there's no way to avoid this extra work in its current form.  Thoughts, @ConnorGray @laszlokindrat?

**Comment by northstreet12** [comment]
Gently ping

**Comment by laszlokindrat** [comment]
> The changes are correct/fine. This may affect performance given how widely variadics are used and there's no way to avoid this extra work in its current form.

Agreed. I'm not sure how big an effect it will actually have, but there is only one way to find out.

**Comment by JoeLoser** [comment]
!sync

**Comment by JoeLoser** [comment]
Do you mind rebasing this PR, please?

**Comment by northstreet12** [comment]
> Do you mind rebasing this PR, please?

OK

**Comment by northstreet12** [comment]
@JoeLoser

**Comment by JoeLoser** [comment]
!sync

**Comment by JoeLoser** [comment]
After some discussion internally, we're not going to accept this change right now.  I appreciate your interest in tackling https://github.com/modularml/mojo/pull/3496, but right now, with our use of runtime variadics internally, this is not the right trade-off (bounds checking with respect to degradation in performance).  Do note that we have a design we're working on to improve the cost (and completely disable if desired) of bounds checking, `debug_assert`, and the like.  Stay tuned for that as it will land soon and then perhaps we could think about this usability friction afterwards.

To help clean up the review queue as I'm working through the many long-standing pull requests that are open, I'm going to close this pull request for now, but we may circle back here on this in the future.

**Pull request closed without merging**